sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala

```diff
@@ -191,11 +195,18 @@ private[sql] class SharedState(
 }
 
 object SharedState extends Logging {
-  try {
-    URL.setURLStreamHandlerFactory(new FsUrlStreamHandlerFactory())
-  } catch {
-    case e: Error =>
-      logWarning("URL.setURLStreamHandlerFactory failed to set FsUrlStreamHandlerFactory")
+  private val fsUrlStreamHandlerFactoryInitialized = new AtomicBoolean(false)
+
+  private def setFsUrlStreamHandlerFactory(conf: SparkConf): Unit = {
+    if (conf.get(DEFAULT_URL_STREAM_HANDLER_FACTORY_ENABLED) &&
+        fsUrlStreamHandlerFactoryInitialized.compareAndSet(false, true)) {
```

Comment:
`getAndSet`?

Reply:
The default value is `false`. If we use `getAndSet`, we need to use `!fsUrlStreamHandlerFactoryInitialized.getAndSet`.

`compareAndSet(false, true)` is the same in this context, isn't it?

Reply:
It returns whether it was successful in changing the value, not a value of the flag. It happens to work here if you invert the comparison, yes. But isn't it a little more confusing as the condition looks like it reads " if(... and initialized) then initialize"

Reply:
Got it. I'll update it for readability.

Reply:
Ur, wait. Let me think about the semantic once more.

Reply:
`getAndSet` is the more straightforward thing to call in any event. You are trying to get the current value, which is easy to reason about, while atomically ensuring it's true going forward.

Reply:
I updated to `getAndSet`.